Seo, SnippetPreview field types

gor0utine opened the pull request.

gor0utine committed:
commit testing revision

gor0utine committed:
remove test.php

gor0utine committed:
add 'Seo' to $arrFields

gor0utine committed:
add 'Seo' field type files

gor0utine committed:
add formSeo() implementation

gor0utine committed:
add 'charslimit' to fieldOptions

gor0utine committed:
sql patch to seo field type

gor0utine committed:
add countChars() method

gor0utine committed:
add seo field textarea counting chars

gor0utine committed:
add seoname, charslimit keys

gor0utine committed:
remove $(elementId) exists checking

gor0utine committed:
add counter to text field

gor0utine committed:
add plus, minus css classes

gor0utine committed:
remove unused code

gor0utine committed:
remove arguments from $this->_getInfo()

gor0utine committed:
add snippet preview sql patch

gor0utine committed:
add snippet preview field type files

gor0utine committed:
add SnippetPreview to $this->arrFields

gor0utine committed:
remove unused 'data' folder in Seo field

gor0utine committed:
add goSnippetPreviewJS()

gor0utine committed:
formSnippetPreview() implementation

gor0utine committed:
add init snippet preview js

gor0utine committed:
set element id to "Test_Page" link

gor0utine committed:
add page url to "snippet preview" field type

gor0utine committed:
set width to .seo_desc

chirimoya commented:
@NurbekChymbaev I will review the PR as soon as possible. THX so fare! /cc @rainerSchoenherr

gor0utine committed:
move seo js functions as myForm methods
add title get variants

gor0utine committed:
add myForm.cleanText()

gor0utine committed:
add myForm.pickOutSeoKeywords()

gor0utine committed:
set .seo_desc 500px

gor0utine committed:
add this.pickOutSeoKeywords() when update #seo_url

gor0utine committed:
increase seo description to 170 chars

gor0utine committed:
sql patch for meta robots

gor0utine committed:
add PageHelper::getMetaRobots() version 1.0

gor0utine committed:
add get_meta_robots()
call get_meta_robots() in master.php

gor0utine committed:
update seo fields on keywords changes
add "onload" argument to myForm.updateSnippetPreviewUrl()

danrot commented:
@NurbekChymbaev I am trying to get your changes running. But I get the following error message when opening a page:

```
exception 'Exception' with message 'Action Helper by name seo not found' in /home/daniel/Workspaces/PHPStorm/ZOOLU/library/massiveart/generic/elements/generic.element.abstract.class.php:167
Stack trace:
#0 /home/daniel/Workspaces/PHPStorm/ZOOLU/library/massiveart/generic/elements/generic.element.abstract.class.php(145): GenericElementAbstract->getHelper()
#1 /home/daniel/Workspaces/PHPStorm/ZOOLU/library/massiveart/generic/data/types/generic.data.type.page.class.php(255): GenericElementAbstract->__call('load', Array)
#2 /home/daniel/Workspaces/PHPStorm/ZOOLU/library/massiveart/generic/data/types/generic.data.type.page.class.php(255): GenericElementField->load('2', 'page', '50699bf79e358', '1')
#3 /home/daniel/Workspaces/PHPStorm/ZOOLU/library/massiveart/generic/forms/generic.form.class.php(205): GenericDataTypePage->load()
#4 /home/daniel/Workspaces/PHPStorm/ZOOLU/application/zoolu/modules/cms/controllers/PageController.php(269): GenericForm->loadFormData()
#5 /opt/lib/ZendFramework-1.12.0/library/Zend/Controller/Action.php(516): Cms_PageController->geteditformAction()
#6 /opt/lib/ZendFramework-1.12.0/library/Zend/Controller/Dispatcher/Standard.php(308): Zend_Controller_Action->dispatch('geteditformActi...')
#7 /opt/lib/ZendFramework-1.12.0/library/Zend/Controller/Front.php(954): Zend_Controller_Dispatcher_Standard->dispatch(Object(Zend_Controller_Request_Http), Object(Zend_Controller_Response_Http))
#8 /home/daniel/Workspaces/PHPStorm/ZOOLU/public/index.php(142): Zend_Controller_Front->dispatch()
#9 {main}
```

I think the action helper for your seo-field is missing, maybe you have only forgotten to commit it?

gor0utine commented:
Yes, I commit all changes. Did you run sql patches for Seo fieldtype?

https://github.com/NurbekChymbaev/ZOOLU/blob/master/installation/db/upgrade/20120920_seo_fieldtype.sql

https://github.com/NurbekChymbaev/ZOOLU/blob/master/installation/db/upgrade/20120922_snippetpreview_fieldtype.sql

https://github.com/NurbekChymbaev/ZOOLU/blob/master/installation/db/upgrade/20120926_seo_meta_robots.sql

danrot commented:
Yes I did... In which file do you handle the storage of these fields?

And BTW, I had some problems with the SQL-Files. MySQL table names are case sensitive on UNIX-based systems. It would be great if you can modify the sql patches, so that they also match the case-sensitive names.

gor0utine commented:
I'm not changed the storage process, it is same as it was in seo_title, seo_description, seo_canonical fields.

I changed only client side: add chars counter, keywords allocation etc.

danrot commented:
Well, is it possible that you have messed up the content of the file 20120920_seo_fieldtype.sql? You insert the value '4' for the idFieldTypeGroup in the first line, which usually means that you will implement the storage process on your own.

gor0utine commented:
Yes, you're right. In my local site I'm set up idFieldTypeGroup as 5. I'll fix.

Sorry.

gor0utine committed:
fix idFieldTypeGroup id

gor0utine commented:
Please, change manually value of idFieldTypeGroup. I hope, now will work correctly.

gor0utine committed:
Merge branch 'master' of https://github.com/massiveart-webservices/ZOOLU

danrot commented:
It's almost working :) Another error is the idDecorator-Value for your seo-fieldType, you have written '0', but I suppose you also want it to be '1'. Can you please also correct this?

Afterwards it runs, but I have seen one little error immediately: In the snippet preview the host part of the address is concanetad at the beginning (e.g in my case: zoolu.localhttp://zoolu.local/de/).

gor0utine commented:
Yes, idDecorator is '1' :( I'll correct.

> Afterwards it runs, but I have seen one little error immediately: In the snippet preview the host part of the address is > concanetad at the beginning (e.g in my case: zoolu.localhttp://zoolu.local/de/).

Yes, this bug appears after last updating zoolu code, Thomas changed #page_url value.

gor0utine committed:
set idDecorator as 1

danrot commented:
But then you can simply set the #page_url value as this green link, or am I missing a point?

We have also encountered a few other things:
- When there are 0 characters left, the 0 should stay green, as you have not exceeded the character limit
- If the user do not set any values you should display the standard values, as it has been before (correct me if this is not true)
- You have used 1 as languageId every time, but this is german. Can you change it to 2 (for english), and we will add the german titles later
- And can you change the sql scripts, so that they match the case-sensitive table names?

gor0utine committed:
fix mysql table case-sensitivity

gor0utine commented:
> When there are 0 characters left, the 0 should stay green, as you have not exceeded the character limit

yes, i'll fix.

> If the user do not set any values you should display the standard values, as it has been before (correct me if this is not true)

yes, it works.

> You have used 1 as languageId every time, but this is german. Can you change it to 2 (for english), and we will add the german titles later

Yes, I'm not added English titles. I can add them, if you approve my changes.

> And can you change the sql scripts, so that they match the case-sensitive table names?

done.

gor0utine committed:
set green text color if count equals 0

danrot commented:
Great, this is working so far, will the other points follow? And why do you need an approval to correct the idLanguages-field in the sql patches?

gor0utine commented:
> Great, this is working so far, will the other points follow? 

Yes, I'm tried to implement next point from ideas and concept

> Duplicate Content Warning: Showing a warning if a page with the same Meta-Title, Description,
> or Keywords already exists

But can't load in code all pages field values, except current page Id. In general, I have specific questions about the code of the Zoolu CMS.

> And why do you need an approval to correct the idLanguages-field in the sql patches?

I was not sure, need to add English titles or not by default when you create new field type. Because Zoolu hasn't everywhere English titles and keywords.

danrot commented:
I meant the points I have posted before: The bug with the wrong link in the snippet preview still exists, and there are still no meta tags on the website when the seo tab is left empty.

And you are right, with the page-object delivered you will only receive the values of the current page. I think you have to look into the Pages-Model, if there are already methods usable for you, or write them on your own. Maybe @chirimoya has a better idea? And what are your specific questions?

I suppose that you are running ZOOLU in english, so it's no problem to write only the english titles into the database. I will add the correct german ones later.

danrot commented:
It would be great if you can first complete these open bug, and do the further work on a seperate branch, so that we can integrate the first part into our repository.

gor0utine commented:
> I meant the points I have posted before: The bug with the wrong link in the snippet preview still exists

Ok. But I have some trouble with new "Seite testen" link on local site, it follows to "http://zoolu.area51.at/zoolu/cms/:///de/was-wir-tun-1/". And because of this, I can't get right page url from "href" of "Seiten testen" link.

> and there are still no meta tags on the website when the seo tab is left empty.

I thought, that we don't need show meta tags(nofollow, noindex) on page front-end, if they are not set on the page editing.

> And you are right, with the page-object delivered you will only receive the values of the current page. I think you have to look into the Pages-Model, if there are already methods usable for you, or write them on your own. Maybe @chirimoya has a better idea? And what are your specific questions?

I tried to use Model_Pages::loadAllPublicPages() to get pages, but it returns all pages include current page.

How we can get current page id from field form helper?

gor0utine commented:
> and there are still no meta tags on the website when the seo tab is left empty.

I've idea to add third parameter "None" to meta(nofollow, noindex) fields, to disable these meta fields on page.

What do you think?

danrot commented:
Strange, the "Seite testen"-Link links to the correct page on my ZOOLU-Instance...

I think that is correct for nofollow and noindex, but there should be at least the meta description, as you also show it on the snippet preview. When I remember correctly it was implemented in the old version of the seo tab.

Isn't there a getId() method in the Page-Object?

And to your ide with the third parameter: I am no seo specialist, but I will ask the creator of the idea & concept document in the office on monday.

danrot commented:
I just asked the creator of the idea and concept document: You should use checkboxes instead of radiobuttons. That means for the robot settings you have 2 checkboxes - noindex and nofollow (index and follow is the default, which means the checkboxes are unchecked). So you don't have to think about a third parameter "nothing".

To summarize, you have to do the following open tasks:
- Replace the radio buttons with checkboxes
- Write the meta tags to the website with default values, if there is no input from the seo-tab
- Show the correct link in the snippet preview (just use the value of the "Seite testen"-Link, I will check if I get the correct value)
- Limit the Meta keywords to 5 words, seperated by comma, instead of the character limitation (this one is new)

Please fulfill these tasks and commit the changes before starting with any new features. Afterwards you can continue working on a new branch, while we will integrate the first version of your work.

gor0utine committed:
change field type for meta fields

gor0utine committed:
get seo url from #page_url

gor0utine committed:
change seo_metaindex, seo_metafollow fieldtypes

gor0utine committed:
set by default values for robots fields

gor0utine commented:
> Limit the Meta keywords to 5 words, seperated by comma, instead of the character limitation (this one is new)

For that, I guess, we need new field type? Called for example as "SeoKeywords".

gor0utine commented:
> Replace the radio buttons with checkboxes
> Write the meta tags to the website with default values, if there is no input from the seo-tab
> Show the correct link in the snippet preview (just use the value of the "Seite testen"-Link, I will check if I get the correct value)

it's done.

gor0utine commented:
There might be problems with new seo meta SQL patch:

https://github.com/NurbekChymbaev/ZOOLU/blob/master/installation/db/upgrade/20120926_seo_meta_robots.sql

It runs correctly only if you doesn't installed `seo_metaindex` and `seo_metafollow` fields. **So don't run it, if you installed before these fields.**

Run next queries, if you have installed these fields:

<pre>
UPDATE `fields` SET `idFieldTypes` = '3' WHERE `name` = 'seo_metaindex' OR `name` = 'seo_metafollow';
UPDATE `fields` SET `sqlSelect` = 'SELECT tbl.id AS id, categoryTitles.title AS title FROM categories AS tbl            INNER JOIN categoryTitles ON categoryTitles.idCategories = tbl.id AND categoryTitles.idLanguages = %LANGUAGE_ID%, categories AS rootCat WHERE rootCat.id = 666 AND tbl.idRootCategory = rootCat.idRootCategory AND tbl.lft BETWEEN ( rootCat.lft +1 ) AND rootCat.rgt %WHERE_ADDON% ORDER BY tbl.lft, categoryTitles.title' WHERE `name` = 'seo_metaindex';
UPDATE `fields` SET `sqlSelect` = 'SELECT tbl.id AS id, categoryTitles.title AS title FROM categories AS tbl INNER JOIN categoryTitles ON categoryTitles.idCategories = tbl.id AND categoryTitles.idLanguages = %LANGUAGE_ID%, categories AS rootCat WHERE rootCat.id = 669 AND tbl.idRootCategory = rootCat.idRootCategory AND tbl.lft BETWEEN ( rootCat.lft +1 ) AND rootCat.rgt %WHERE_ADDON% ORDER BY tbl.lft, categoryTitles.title' WHERE `name` = 'seo_metafollow';
DELETE FROM categoryTitles WHERE `title` = 'index' OR `title` = 'follow';
DELETE FROM `categories` WHERE `categories`.`id` = 668;
DELETE FROM `categories` WHERE `categories`.`id` = 671;
UPDATE categories SET `lft` = '1', `rgt` = '4' WHERE `id` = 666;
UPDATE categories SET `lft` = '2', `rgt` = '3' WHERE `id` = 667;
UPDATE categories SET `lft` = '1', `rgt` = '4' WHERE `id` = 669;
UPDATE categories SET `lft` = '2', `rgt` = '3' WHERE `id` = 670;
UPDATE categoryTitles SET idCategories=667 WHERE `title` = 'noindex';
UPDATE categoryTitles SET idCategories=670 WHERE `title` = 'nofollow';
ALTER TABLE `page-DEFAULT_PAGE_1-1-Instances`
  DROP `seo_metaindex`,
  DROP `seo_metafollow`;
</pre>

danrot commented:
To the point with the seo keyword limitation: Maybe the options-field in the fields-table can be used to differ between description and keywords, this way you possibly can avoid code duplication. But I am not sure if it's read by ZOOLU... If not you are correct, you would have to create a new fieldType.

And I will have a look at the other changes as soon as possible.

danrot commented:
- Replace the radio buttons with checkboxes
  Works partially, there is a foreign key constraint failing when inserting the category title with the id 670. Next you should have one multicheckbox, which means there is a heading "Meta Robots" with two checkboxes saying "noindex" and "nofollow". And are you sure have understood the category-table correctly? May be you should set these values under the "Einstellungen"-Link in the right top of ZOOLU, and then choose "System Interne" on the left.
- Write the meta tags to the website with default values, if there is no input from the seo-tab
  This one is ok
- Show the correct link in the snippet preview (just use the value of the "Seite testen"-Link, I will check if I get the correct value)
  I now see the correct value, but I have no idea why it's not working on your machine...
- Limit the Meta keywords to 5 words, seperated by comma, instead of the character limitation (this one is new)
  See my previous comment

gor0utine commented:
> Replace the radio buttons with checkboxes
> Works partially, there is a foreign key constraint failing when inserting the category title with the id 670. Next you should have one multicheckbox, which means there is a heading "Meta Robots" with two checkboxes saying "noindex" and "nofollow".

Ok, I'll fix.

> And are you sure have understood the category-table correctly? May be you should set these values under the "Einstellungen"-Link in the right top of ZOOLU, and then choose "System Interne" on the left.

No, I'm not sure, me need figure out how works "Einstellungen" part.

gor0utine committed:
union follow, index fields to one

gor0utine commented:
> Replace the radio buttons with checkboxes
> Works partially, there is a foreign key constraint failing when inserting the category title with the id 670. Next you should have one multicheckbox, which means there is a heading "Meta Robots" with two checkboxes saying "noindex" and "nofollow".

So, it's done. Use next queries to update DB.

<pre>
DELETE FROM `fields` WHERE `name` = 'seo_metafollow';
UPDATE `fields` SET `name` = 'seo_metarobots' WHERE `name` = 'seo_metaindex';
DELETE FROM `fieldTitles` WHERE `title` = 'Meta Robots Follow';
UPDATE `fieldTitles` SET `title` = 'Meta Robots' WHERE `title` = 'Meta Robots Index';
DELETE FROM `regionFields` WHERE `idRegions` = 99 AND `idFields` = 264;
</pre>

> And are you sure have understood the category-table correctly? May be you should set these values under the "Einstellungen"-Link in the right top of ZOOLU, and then choose "System Interne" on the left.

So, I figured out how works "Einstellungen" part.

First I deleted all old data from `categories`, `categoryTitles` tables. Using next sql:
<code>
DELETE FROM `categories` WHERE `id` >= 666 AND `id` <= 670;
DELETE FROM `categoryTitles` WHERE `title` = 'noindex' OR `title` = 'nofollow';
</code>

Then I added in "Einstellungen" -> "System Interne", "Meta Robots" category, and added to it "nofollow", "noindex" subcategories.

gor0utine committed:
set rootCat.id = 676

gor0utine committed:
get robots from union index, follow fields

gor0utine commented:
How we can make universal SQL query for `seo_metarobots` field ? Current is:

<pre>
SELECT tbl.id AS id, categoryTitles.title AS title 
     FROM categories AS tbl 
     INNER JOIN categoryTitles ON categoryTitles.idCategories = tbl.id 
               AND categoryTitles.idLanguages = %LANGUAGE_ID%, categories AS rootCat
  WHERE 
              rootCat.id = 676 
      AND tbl.idRootCategory = rootCat.idRootCategory 
      AND tbl.lft BETWEEN ( rootCat.lft +1 ) 
      AND rootCat.rgt %WHERE_ADDON% 
  ORDER BY tbl.lft, categoryTitles.title
</pre>

It uses `rootCat.id = 676` in my database, but in other computers it can be different.

The results of query depend on category id, that can be different.

Any ideas?

danrot commented:
It is working, there is one notice shown for each of the two checkbox not checked:

```
( ! ) Notice: Uninitialized string offset: 0 in /home/daniel/Workspaces/PHPStorm/ZOOLU/application/website/default/helpers/PageHelper.php on line 415
Call Stack
#   Time    Memory  Function    Location
1   0.0010  257040  {main}( )   ../index.php:0
2   0.0701  6185328 Zend_Controller_Front->dispatch( )  ../index.php:142
3   0.0721  6489320 Zend_Controller_Dispatcher_Standard->dispatch( )    ../Front.php:954
4   0.0766  7085400 Zend_Controller_Action->dispatch( ) ../Standard.php:308
5   0.0766  7096072 IndexController->indexAction( ) ../Action.php:516
6   0.2384  14215688    Zend_Controller_Action->renderScript( ) ../IndexController.php:544
7   0.2385  14216432    Zend_Controller_Action_Helper_ViewRenderer->renderScript( ) ../Action.php:243
8   0.2385  14216640    Zend_View_Abstract->render( )   ../ViewRenderer.php:900
9   0.2385  14233688    Zend_View->_run( )  ../Abstract.php:888
10  0.2386  14247632    include( '/home/daniel/Workspaces/PHPStorm/ZOOLU/public/website/themes/default/master.php' )    ../View.php:108
11  0.2387  14248768    get_meta_robots( )  ../master.php:6
12  0.2387  14248856    PageHelper->getMetaRobots( )    ../page.inc.php:108

( ! ) Notice: Uninitialized string offset: 1 in /home/daniel/Workspaces/PHPStorm/ZOOLU/application/website/default/helpers/PageHelper.php on line 424
Call Stack
#   Time    Memory  Function    Location
1   0.0010  257040  {main}( )   ../index.php:0
2   0.0701  6185328 Zend_Controller_Front->dispatch( )  ../index.php:142
3   0.0721  6489320 Zend_Controller_Dispatcher_Standard->dispatch( )    ../Front.php:954
4   0.0766  7085400 Zend_Controller_Action->dispatch( ) ../Standard.php:308
5   0.0766  7096072 IndexController->indexAction( ) ../Action.php:516
6   0.2384  14215688    Zend_Controller_Action->renderScript( ) ../IndexController.php:544
7   0.2385  14216432    Zend_Controller_Action_Helper_ViewRenderer->renderScript( ) ../Action.php:243
8   0.2385  14216640    Zend_View_Abstract->render( )   ../ViewRenderer.php:900
9   0.2385  14233688    Zend_View->_run( )  ../Abstract.php:888
10  0.2386  14247632    include( '/home/daniel/Workspaces/PHPStorm/ZOOLU/public/website/themes/default/master.php' )    ../View.php:108
11  0.2387  14248768    get_meta_robots( )  ../master.php:6
12  0.2387  14248856    PageHelper->getMetaRobots( )    ../page.inc.php:108
```

Don't you have notices activated in your php.ini? I would recommend it to you!

And you are right with the sql query, it's something which had not be thought enough about at the beginning of ZOOLU. We will have to decide the correct ids at the end of you development, and write them in the mysql dump delivered with ZOOLU.

gor0utine committed:
fix notices

gor0utine committed:
update tables by `name` instead `id`

gor0utine committed:
add 'SeoKeywords' to Zend_Loader_PluginLoader::arrFields

gor0utine committed:
add limitKeywords()

gor0utine committed:
SeoKeywords field type files

gor0utine committed:
SeoKeywords field type SQL patch
update seo_keywords field

gor0utine committed:
new SeoKeywords validator

gor0utine commented:
> Limit the Meta keywords to 5 words, seperated by comma, instead of the character limitation (this one is new)

I did this functionality. Create new field type which is called as `SeoKeywords`. To verify keywords limit in client side, added new function to `form.js`.

To verify in backend use new field validator `SeoKeywords`.

Run new SQL patch.

How are you add language messages? I added new one using `$this->core->translate->_('Err_extra_keywords')`.

danrot commented:
Is this the reason for the missing text below the keyword field? And you can translate this message using the little program `poedit`. The files for this program are in /application/zoolu/language. Consider that you have to set the correct paths in poedit, and be sure that you have set it up to automatically create .mo-files.

gor0utine committed:
add German translate to 'Err_extra_keywords' message

gor0utine commented:
Translation is done. What is next?

danrot commented:
Ok, I'll check that tomorrow, and you can try to rename your branch according to our new guidelines: https://github.com/massiveart-webservices/ZOOLU/wiki/Branching-model

Please also send a new pull request, with our develop-branch as target.

Afterwards you can start the other points on a separate branch.

gor0utine commented:
Me need to rename `master` branch? Or need to create new branch `feature/seo-tab-extension` from `master`?

I'm not experienced git user. Please, write details.

Thanks.

danrot commented:
I hope that you can rename the master branch, make a backup and try it. When you have done it, send the new pull request, as I have already described.

Afterwards create a new branch from you `feature/seo-tab-extension`, in which you can continue your development. If I find another error, you can easily switch to the old branch, and fix them.
Otherwise your new work would be added to this pull request, and we would not be able to merge the changes again.

And I still don't see any text with the keyword count below the keyword textarea. Is it working on your machine?

gor0utine commented:
> And I still don't see any text with the keyword count below the keyword textarea. Is it working on your machine?

It's not implemented. I'll do.

gor0utine committed:
show keywords count under seo keywords

gor0utine commented:
> I hope that you can rename the master branch, make a backup and try it. When you have done it, send the new pull request, as I have already described.

Sorry, I could not to rename my `master` branch.

> Afterwards create a new branch from you feature/seo-tab-extension, in which you can continue your development. If I find another error, you can easily switch to the old branch, and fix them.
> Otherwise your new work would be added to this pull request, and we would not be able to merge the changes again.

I added `feature/seo-tab-extension` branch, and send Pull Request.

danrot commented:
Well, the new branch seems to work, but you have chosen the wrong target branch.
Can you close these two pull requests, and send a new one from with your feature/seo-tab-extension as head branch, and our develop branch as the base branch.

And to the keyword field: Can't you do it the same way as the description? I mean that the keyword counter can also increase higher than 5, but it should get red then.

gor0utine commented:
> And to the keyword field: Can't you do it the same way as the description? I mean that the keyword counter can also increase higher than 5, but it should get red then.

Ok, done. Do we need verify extra keywords on server side?

gor0utine commented:
> Well, the new branch seems to work, but you have chosen the wrong target branch.
> Can you close these two pull requests, and send a new one from with your feature/seo-tab-extension as head branch, and our develop branch as the base branch.

done

gor0utine commented:
https://github.com/massiveart-webservices/ZOOLU/pull/7
